core/shared/src/main/scala/zio/internal/FiberRuntime.scala

```diff
@@ -457,7 +455,10 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
 
       val supervisor = getSupervisor()
 
-      if (supervisor ne Supervisor.none) supervisor.onSuspend(self)(Unsafe)
+      if (supervisor ne Supervisor.none) {
+        if (finalExit ne null) supervisor.onEnd(finalExit, self)(Unsafe)
+        supervisor.onSuspend(self)(Unsafe)
```

Comment:
I think there should probably be an `else` statement here, but I'm worried this might cause issues in cases we're expecting the `onSuspend` method to be called regardless whether we've suspended or exited so I decided to leave it as it was (with `onEnd` being called first as that was the behaviour previously)